cmd/up.go

```diff
@@ -819,3 +840,222 @@ func setupDotfiles(
+	gpgTrustImport = append(gpgTrustImport, "gpg --import-ownertrust")
+	gpgTrustImportCmd := exec.Command(execPath, gpgTrustImport...)
+
+	stdin, err = gpgTrustImportCmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer stdin.Close()
+		_, _ = stdin.Write(ownerTrustExport)
+	}()
+
+	log.Debugf("gpg: importing gpg owner trust in container")
+
+	err = gpgTrustImportCmd.Run()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("gpg: detecting gpg-agent socket path on host")
+	// Detect local agent extra socket, this will be forwarded to the remote and
+	// symlinked in multiple paths
+	gpgExtraSocketBytes, err := exec.Command("gpgconf", []string{"--list-dir", "agent-extra-socket"}...).
+		Output()
+	if err != nil {
+		return err
+	}
+
+	gpgExtraSocketPath := strings.TrimSpace(string(gpgExtraSocketBytes))
+
+	log.Debugf("gpg: detected gpg-agent socket path %s", gpgExtraSocketPath)
+
+	// Kill the agent in the workspace if it was started by one of the above
+	// commands.
+	killAgent := append(sshCmdArgs, "--command")
+	killAgent = append(
+		killAgent,
+		fmt.Sprintf("gpgconf --kill gpg-agent && rm -f %q", gpgExtraSocketPath),
+	)
+
+	log.Debugf("gpg: ensuring gpg-agent is not running in container")
+
+	err = exec.Command(execPath, killAgent...).Run()
+	if err != nil {
+		return err
+	}
+
+	// Now we forward the agent socket to the remote, and setup remote gpg to use it
+	// fix eventual permissions and so on
+	forwardAgent := append(sshCmdArgs, "--reverse-forward-ports")
+	forwardAgent = append(forwardAgent, gpgExtraSocketPath)
+
+	forwardAgentCmd := exec.Command(execPath, forwardAgent...)
+
+	forwardAgentCmd.Stdout = writer
+	forwardAgentCmd.Stderr = writer
+
+	log.Debugf(
+		"gpg: start reverse forward of gpg-agent socket %s, keeping connection open",
+		gpgExtraSocketPath,
+	)
+
+	// We use start to keep this connection alive in background (hence the use of sleep infinity)
+	err = forwardAgentCmd.Start()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("gpg: fixing agent paths and permissions")
+
+	// Permissions may not be correctly set, or paths can differ between systems
+	// so here we ensure the default user is able to read the sockets and link
+	// the socket to canonical places
+	fixAgent := append(sshCmdArgs, "--command")
+	fixAgent = append(fixAgent, `sh -c '
+set -x
+
+sleep 5
+
+sudo mkdir -p /run/user/$(id -ru)/gnupg
+sudo chmod 0700 /run/user/$(id -ru)/gnupg
+sudo chown -R  $(id -ru):$(id -rg) /run/user
+sudo chown $(id -ru):$(id -rg) `+gpgExtraSocketPath+`
+
+mkdir -p $HOME/.gnupg
+chmod 0700 $HOME/.gnupg
+
+sudo ln -sf `+gpgExtraSocketPath+` /run/user/$(id -ru)/gnupg/S.gpg-agent
+sudo ln -sf `+gpgExtraSocketPath+` $HOME/.gnupg/S.gpg-agent
+
+if ! grep -q "use-agent" "$HOME/.gnupg/gpg.conf"; then
+	echo "use-agent" >> "$HOME/.gnupg/gpg.conf"
+fi
+'`)
+
+	err = exec.Command(execPath, fixAgent...).Run()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("gpg-agent forwarding done")
+
+	return nil
```

Comment:
can we condense all this into a single command that is executed? so the goal should be to only call exec.Command once

Reply:
Done